### ISSUE TITLE
[BUG] INNR SP 240 "summation delivered" kWh value is 100x expected value

### ISSUE THREAD

**Issue by rrozema** [zigpy/zha-device-handlers#2781]
### Bug description

Upon inclusion of the INNR SP 240 in ZHA, among others a "Summation delivered" sensor gets created. The value returned for this sensor is however 100x larger than expected: I ran a filament dryer for 5 hours of the SP 240 and the device measured 17kWh instead of the expected 0,17 kWh as measured using a Fibaro wall plug V2.

When I requested help from INNR support they responded that ZHA probably has incorrectly interpreted the configuration their device returned. My translation of their response: "An attribute "SummationFormatting" on the Metering cluster has the value 0x2A, which designates: 2 digits to the right of the decimal point, 5 digits to the left of the decimal point, don't suppress leading zeroes". Looks like his system (red: my ZHA integration in HA)) doesn't interpret or incorrectly interprets the value of the "SummationFormatting" attribute." 

The full mail text I got from INNR support (in Dutch, sorry):

> Beste Richard,
> 
>  
> 
> (This message is in Dutch because I assume based on your name and e-mail address that you also speak Dutch.)
> 
> Ik heb de vraag even bij één van onze engineers gelegd, en dit was zijn reactie:
> 
> Wat betreft Watt meting die op 0 decimalen wordt afgerond:
> 
> De waarde wordt opgeslagen en gecommuniceerd in een “int16” type, wat een bereik heeft van -32768 tot 32767. Als we op 1 decimaal zouden afronden zou het maximum slechts 3276.7 W zijn, maar de plug kan 3680 W aan dus de keuze was simpel. De afronding dynamisch maken (meer decimalen bij lagere waarden) is geen optie want dat ondersteunen de meeste systemen niet (voor zover ik weet geen enkel systeem).
> 
>  
> 
> Wat betreft de kWh formatting: het “SummationFormatting” attribuut van het Metering cluster heeft waarde 0x2A, wat betekent: 2 digits to the right of the decimal point, 5 digits to the left of the decimal point, don’t suppress leading zeroes. Lijkt er dus op dat zijn systeem de waarde van het “SummationFormatting” attribuut van het Metering cluster niet of verkeerd interpreteert.
> 
>  
> 
> Hopelijk heb je iets aan deze uitleg.
> 
>  
> 
> Met vriendelijke groet,
> 
>  
> 
>  
> 
> **REDACTED**
> 
> Innr Servicedesk 
> Innr Lighting | Innr.com  
>  
> Innr Lighting BV  |  Catharina van Renneslaan 20  |  1217 CX Hilversum
> 

@TheJulianJES requested me to tag him in here.

### Steps to reproduce

Include the INNR SP 240 into the network using ZHA,
plug in a filament dryer of 145W,
Run the dryer for 5 hours, Watt history graph shows a peak at the beginning, then dropping to roughly 30 Watt.
Multiply the measured Watt value with the number of hours: 5 * roughly 30 Watt = roughly 170 Wh, or 0,17 kWh.
See how the actual sensor in ZHA however shows 17kWh.

### Expected behavior

The summation delivered sensor should have shown a usage of 0,170 kWh

### Screenshots/Video

<details><summary>Screenshots/Video</summary>

![image](https://github.com/zigpy/zha-device-handlers/assets/1747982/bdca1488-df6e-4c55-91be-b64dbf4e8422)

![image](https://github.com/zigpy/zha-device-handlers/assets/1747982/93985e4a-2c74-4af4-825a-562a472833da)

![image](https://github.com/zigpy/zha-device-handlers/assets/1747982/02eda6be-7639-49e8-9c98-b55792fab7ef)

</details>

### Device signature

<details><summary>Device signature</summary>

```json
{
  "node_descriptor": "NodeDescriptor(logical_type=<LogicalType.Router: 1>, complex_descriptor_available=0, user_descriptor_available=0, reserved=0, aps_flags=0, frequency_band=<FrequencyBand.Freq2400MHz: 8>, mac_capability_flags=<MACCapabilityFlags.FullFunctionDevice|MainsPowered|RxOnWhenIdle|AllocateAddress: 142>, manufacturer_code=4454, maximum_buffer_size=66, maximum_incoming_transfer_size=66, server_mask=10752, maximum_outgoing_transfer_size=66, descriptor_capability_field=<DescriptorCapability.NONE: 0>, *allocate_address=True, *is_alternate_pan_coordinator=False, *is_coordinator=False, *is_end_device=False, *is_full_function_device=True, *is_mains_powered=True, *is_receiver_on_when_idle=True, *is_router=True, *is_security_capable=False)",
  "endpoints": {
    "1": {
      "profile_id": "0x0104",
      "device_type": "0x010a",
      "input_clusters": [
        "0x0000",
        "0x0003",
        "0x0004",
        "0x0005",
        "0x0006",
        "0x0008",
        "0x0702",
        "0x0b04",
        "0x1000",
        "0xe001"
      ],
      "output_clusters": [
        "0x000a",
        "0x0019"
      ]
    },
    "242": {
      "profile_id": "0xa1e0",
      "device_type": "0x0061",
      "input_clusters": [],
      "output_clusters": [
        "0x0021"
      ]
    }
  },
  "manufacturer": "innr",
  "model": "SP 240",
  "class": "zigpy.device.Device"
}
```

</details>

### Diagnostic information

<details><summary>Diagnostic information</summary>

```json
{
  "home_assistant": {
    "installation_type": "Home Assistant OS",
    "version": "2023.11.3",
    "dev": false,
    "hassio": true,
    "virtualenv": false,
    "python_version": "3.11.6",
    "docker": true,
    "arch": "x86_64",
    "timezone": "Europe/Amsterdam",
    "os_name": "Linux",
    "os_version": "6.1.59",
    "supervisor": "2023.11.3",
    "host_os": "Home Assistant OS 11.1",
    "docker_version": "24.0.6",
    "chassis": "embedded",
    "run_as_root": true
  },
  "custom_components": {
    "battery_consumption": {
      "version": "1.0.4",
      "requirements": []
    },
    "spook": {
      "version": "1.3.2",
      "requirements": []
    },
    "chime_tts": {
      "version": "v0.8.0",
      "requirements": [
        "pydub"
      ]
    },
    "battery_sim": {
      "version": "1.0",
      "requirements": []
    },
    "google_fit": {
      "version": "v2.1.2",
      "requirements": [
        "oauth2client==4.1.3",
        "google-api-python-client==2.82.0"
      ]
    },
    "hacs": {
      "version": "1.33.0",
      "requirements": [
        "aiogithubapi>=22.10.1"
      ]
    },
    "energy_meter": {
      "version": "1.0.1",
      "requirements": []
    },
    "adaptive_lighting": {
      "version": "1.19.0",
      "requirements": [
        "ulid-transform"
      ]
    },
    "smart_irrigation": {
      "version": "0.0.81",
      "requirements": []
    },
    "hass_agent": {
      "version": "2022.11.9",
      "requirements": []
    },
    "var": {
      "version": "0.15.0",
      "requirements": []
    },
    "presence_simulation": {
      "version": "3.2",
      "requirements": []
    },
    "grad_vol": {
      "version": "1.0.0",
      "requirements": []
    },
    "better_thermostat": {
      "version": "1.4.0",
      "requirements": []
    },
    "xiaomi_cloud_map_extractor": {
      "version": "v2.2.0",
      "requirements": [
        "pillow",
        "pybase64",
        "python-miio",
        "requests",
        "pycryptodome"
      ]
    },
    "powercalc": {
      "version": "v1.9.7",
      "requirements": [
        "numpy>=1.21.1"
      ]
    },
    "spook_inverse": {
      "version": "1.3.2",
      "requirements": []
    }
  },
  "integration_manifest": {
    "domain": "zha",
    "name": "Zigbee Home Automation",
    "after_dependencies": [
      "onboarding",
      "usb"
    ],
    "codeowners": [
      "@dmulcahey",
      "@adminiuga",
      "@puddly"
    ],
    "config_flow": true,
    "dependencies": [
      "file_upload"
    ],
    "documentation": "https://www.home-assistant.io/integrations/zha",
    "iot_class": "local_polling",
    "loggers": [
      "aiosqlite",
      "bellows",
      "crccheck",
      "pure_pcapy3",
      "zhaquirks",
      "zigpy",
      "zigpy_deconz",
      "zigpy_xbee",
      "zigpy_zigate",
      "zigpy_znp",
      "universal_silabs_flasher"
    ],
    "requirements": [
      "bellows==0.36.8",
      "pyserial==3.5",
      "pyserial-asyncio==0.6",
      "zha-quirks==0.0.106",
      "zigpy-deconz==0.21.1",
      "zigpy==0.59.0",
      "zigpy-xbee==0.19.0",
      "zigpy-zigate==0.11.0",
      "zigpy-znp==0.11.6",
      "universal-silabs-flasher==0.0.14",
      "pyserial-asyncio-fast==0.11"
    ],
    "usb": [
      {
        "vid": "10C4",
        "pid": "EA60",
        "description": "*2652*",
        "known_devices": [
          "slae.sh cc2652rb stick"
        ]
      },
      {
        "vid": "1A86",
        "pid": "55D4",
        "description": "*sonoff*plus*",
        "known_devices": [
          "sonoff zigbee dongle plus v2"
        ]
      },
      {
        "vid": "10C4",
        "pid": "EA60",
        "description": "*sonoff*plus*",
        "known_devices": [
          "sonoff zigbee dongle plus"
        ]
      },
      {
        "vid": "10C4",
        "pid": "EA60",
        "description": "*tubeszb*",
        "known_devices": [
          "TubesZB Coordinator"
        ]
      },
      {
        "vid": "1A86",
        "pid": "7523",
        "description": "*tubeszb*",
        "known_devices": [
          "TubesZB Coordinator"
        ]
      },
      {
        "vid": "1A86",
        "pid": "7523",
        "description": "*zigstar*",
        "known_devices": [
          "ZigStar Coordinators"
        ]
      },
      {
        "vid": "1CF1",
        "pid": "0030",
        "description": "*conbee*",
        "known_devices": [
          "Conbee II"
        ]
      },
      {
        "vid": "10C4",
        "pid": "8A2A",
        "description": "*zigbee*",
        "known_devices": [
          "Nortek HUSBZB-1"
        ]
      },
      {
        "vid": "0403",
        "pid": "6015",
        "description": "*zigate*",
        "known_devices": [
          "ZiGate+"
        ]
      },
      {
        "vid": "10C4",
        "pid": "EA60",
        "description": "*zigate*",
        "known_devices": [
          "ZiGate"
        ]
      },
      {
        "vid": "10C4",
        "pid": "8B34",
        "description": "*bv 2010/10*",
        "known_devices": [
          "Bitron Video AV2010/10"
        ]
      }
    ],
    "zeroconf": [
      {
        "type": "_esphomelib._tcp.local.",
        "name": "tube*"
      },
      {
        "type": "_zigate-zigbee-gateway._tcp.local.",
        "name": "*zigate*"
      },
      {
        "type": "_zigstar_gw._tcp.local.",
        "name": "*zigstar*"
      },
      {
        "type": "_uzg-01._tcp.local.",
        "name": "uzg-01*"
      },
      {
        "type": "_slzb-06._tcp.local.",
        "name": "slzb-06*"
      }
    ],
    "is_built_in": true
  },
  "data": {
    "ieee": "**REDACTED**",
    "nwk": 12957,
    "manufacturer": "innr",
    "model": "SP 240",
    "name": "innr SP 240",
    "quirk_applied": false,
    "quirk_class": "zigpy.device.Device",
    "quirk_id": null,
    "manufacturer_code": 4454,
    "power_source": "Mains",
    "lqi": 123,
    "rssi": null,
    "last_seen": "2023-11-26T21:32:06",
    "available": true,
    "device_type": "Router",
    "signature": {
      "node_descriptor": "NodeDescriptor(logical_type=<LogicalType.Router: 1>, complex_descriptor_available=0, user_descriptor_available=0, reserved=0, aps_flags=0, frequency_band=<FrequencyBand.Freq2400MHz: 8>, mac_capability_flags=<MACCapabilityFlags.FullFunctionDevice|MainsPowered|RxOnWhenIdle|AllocateAddress: 142>, manufacturer_code=4454, maximum_buffer_size=66, maximum_incoming_transfer_size=66, server_mask=10752, maximum_outgoing_transfer_size=66, descriptor_capability_field=<DescriptorCapability.NONE: 0>, *allocate_address=True, *is_alternate_pan_coordinator=False, *is_coordinator=False, *is_end_device=False, *is_full_function_device=True, *is_mains_powered=True, *is_receiver_on_when_idle=True, *is_router=True, *is_security_capable=False)",
      "endpoints": {
        "1": {
          "profile_id": "0x0104",
          "device_type": "0x010a",
          "input_clusters": [
            "0x0000",
            "0x0003",
            "0x0004",
            "0x0005",
            "0x0006",
            "0x0008",
            "0x0702",
            "0x0b04",
            "0x1000",
            "0xe001"
          ],
          "output_clusters": [
            "0x000a",
            "0x0019"
          ]
        },
        "242": {
          "profile_id": "0xa1e0",
          "device_type": "0x0061",
          "input_clusters": [],
          "output_clusters": [
            "0x0021"
          ]
        }
      },
      "manufacturer": "innr",
      "model": "SP 240"
    },
    "active_coordinator": false,
    "entities": [
      {
        "entity_id": "button.woonkamer_innr_sp240_identify",
        "name": "innr SP 240"
      },
      {
        "entity_id": "select.woonkamer_innr_sp240_start_up_behavior",
        "name": "innr SP 240"
      },
      {
        "entity_id": "sensor.woonkamer_innr_sp240_current",
        "name": "innr SP 240"
      },
      {
        "entity_id": "sensor.woonkamer_innr_sp240_voltage",
        "name": "innr SP 240"
      },
      {
        "entity_id": "sensor.woonkamer_innr_sp240_electric_consumption_w",
        "name": "innr SP 240"
      },
      {
        "entity_id": "sensor.woonkamer_innr_sp240_electric_consumption_kwh",
        "name": "innr SP 240"
      },
      {
        "entity_id": "switch.woonkamer_kerstverlichting_3",
        "name": "innr SP 240"
      }
    ],
    "neighbors": [
      {
        "device_type": "Coordinator",
        "rx_on_when_idle": "On",
        "relationship": "Parent",
        "extended_pan_id": "**REDACTED**",
        "ieee": "**REDACTED**",
        "nwk": "0x0000",
        "permit_joining": "Unknown",
        "depth": "0",
        "lqi": "48"
      },
      {
        "device_type": "Router",
        "rx_on_when_idle": "On",
        "relationship": "Sibling",
        "extended_pan_id": "**REDACTED**",
        "ieee": "**REDACTED**",
        "nwk": "0xA795",
        "permit_joining": "Unknown",
        "depth": "0",
        "lqi": "62"
      },
      {
        "device_type": "Router",
        "rx_on_when_idle": "On",
        "relationship": "Sibling",
        "extended_pan_id": "**REDACTED**",
        "ieee": "**REDACTED**",
        "nwk": "0xF80E",
        "permit_joining": "Unknown",
        "depth": "0",
        "lqi": "223"
      },
      {
        "device_type": "Router",
        "rx_on_when_idle": "On",
        "relationship": "Sibling",
        "extended_pan_id": "**REDACTED**",
        "ieee": "**REDACTED**",
        "nwk": "0x0316",
        "permit_joining": "Unknown",
        "depth": "0",
        "lqi": "88"
      },
      {
        "device_type": "Router",
        "rx_on_when_idle": "On",
        "relationship": "Sibling",
        "extended_pan_id": "**REDACTED**",
        "ieee": "**REDACTED**",
        "nwk": "0xFF73",
        "permit_joining": "Unknown",
        "depth": "0",
        "lqi": "80"
      },
      {
        "device_type": "Router",
        "rx_on_when_idle": "On",
        "relationship": "Sibling",
        "extended_pan_id": "**REDACTED**",
        "ieee": "**REDACTED**",
        "nwk": "0x2415",
        "permit_joining": "Unknown",
        "depth": "0",
        "lqi": "45"
      },
      {
        "device_type": "Router",
        "rx_on_when_idle": "On",
        "relationship": "Sibling",
        "extended_pan_id": "**REDACTED**",
        "ieee": "**REDACTED**",
        "nwk": "0xA16F",
        "permit_joining": "Unknown",
        "depth": "0",
        "lqi": "0"
      },
      {
        "device_type": "Router",
        "rx_on_when_idle": "On",
        "relationship": "Sibling",
        "extended_pan_id": "**REDACTED**",
        "ieee": "**REDACTED**",
        "nwk": "0x6D36",
        "permit_joining": "Unknown",
        "depth": "0",
        "lqi": "53"
      },
      {
        "device_type": "Router",
        "rx_on_when_idle": "On",
        "relationship": "Sibling",
        "extended_pan_id": "**REDACTED**",
        "ieee": "**REDACTED**",
        "nwk": "0xF373",
        "permit_joining": "Unknown",
        "depth": "0",
        "lqi": "103"
      },
      {
        "device_type": "Router",
        "rx_on_when_idle": "On",
        "relationship": "Sibling",
        "extended_pan_id": "**REDACTED**",
        "ieee": "**REDACTED**",
        "nwk": "0xF608",
        "permit_joining": "Unknown",
        "depth": "0",
        "lqi": "47"
      },
      {
        "device_type": "Router",
        "rx_on_when_idle": "On",
        "relationship": "Sibling",
        "extended_pan_id": "**REDACTED**",
        "ieee": "**REDACTED**",
        "nwk": "0xC07F",
        "permit_joining": "Unknown",
        "depth": "0",
        "lqi": "118"
      },
      {
        "device_type": "Router",
        "rx_on_when_idle": "On",
        "relationship": "Sibling",
        "extended_pan_id": "**REDACTED**",
        "ieee": "**REDACTED**",
        "nwk": "0xB58E",
        "permit_joining": "Unknown",
        "depth": "0",
        "lqi": "16"
      },
      {
        "device_type": "Router",
        "rx_on_when_idle": "On",
        "relationship": "Sibling",
        "extended_pan_id": "**REDACTED**",
        "ieee": "**REDACTED**",
        "nwk": "0x3242",
        "permit_joining": "Unknown",
        "depth": "0",
        "lqi": "66"
      },
      {
        "device_type": "Router",
        "rx_on_when_idle": "On",
        "relationship": "Sibling",
        "extended_pan_id": "**REDACTED**",
        "ieee": "**REDACTED**",
        "nwk": "0xA1BB",
        "permit_joining": "Unknown",
        "depth": "0",
        "lqi": "29"
      },
      {
        "device_type": "Router",
        "rx_on_when_idle": "On",
        "relationship": "Sibling",
        "extended_pan_id": "**REDACTED**",
        "ieee": "**REDACTED**",
        "nwk": "0x8A77",
        "permit_joining": "Unknown",
        "depth": "0",
        "lqi": "15"
      }
    ],
    "routes": [],
    "endpoint_names": [
      {
        "name": "ON_OFF_PLUG_IN_UNIT"
      },
      {
        "name": "PROXY_BASIC"
      }
    ],
    "user_given_name": "Woonkamer: innr sp240",
    "device_reg_id": "7fa8dcff82a4c9d75905608870c6f7ef",
    "area_id": "woonkamer",
    "cluster_details": {
      "1": {
        "device_type": {
          "name": "ON_OFF_PLUG_IN_UNIT",
          "id": 266
        },
        "profile_id": 260,
        "in_clusters": {
          "0x0000": {
            "endpoint_attribute": "basic",
            "attributes": {
              "0x0004": {
                "attribute_name": "manufacturer",
                "value": "innr"
              },
              "0x0005": {
                "attribute_name": "model",
                "value": "SP 240"
              }
            },
            "unsupported_attributes": {}
          },
          "0x0003": {
            "endpoint_attribute": "identify",
            "attributes": {},
            "unsupported_attributes": {}
          },
          "0x0004": {
            "endpoint_attribute": "groups",
            "attributes": {},
            "unsupported_attributes": {}
          },
          "0x0005": {
            "endpoint_attribute": "scenes",
            "attributes": {},
            "unsupported_attributes": {}
          },
          "0x0006": {
            "endpoint_attribute": "on_off",
            "attributes": {
              "0x0000": {
                "attribute_name": "on_off",
                "value": 1
              },
              "0x4003": {
                "attribute_name": "start_up_on_off",
                "value": 255
              }
            },
            "unsupported_attributes": {}
          },
          "0x0008": {
            "endpoint_attribute": "level",
            "attributes": {
              "0x0000": {
                "attribute_name": "current_level",
                "value": 1
              }
            },
            "unsupported_attributes": {
              "0x4000": {
                "attribute_name": "start_up_current_level"
              },
              "0x0012": {
                "attribute_name": "on_transition_time"
              },
              "0x0010": {
                "attribute_name": "on_off_transition_time"
              },
              "0x0011": {
                "attribute_name": "on_level"
              },
              "0x0013": {
                "attribute_name": "off_transition_time"
              },
              "0x0014": {
                "attribute_name": "default_move_rate"
              }
            }
          },
          "0x1000": {
            "endpoint_attribute": "lightlink",
            "attributes": {},
            "unsupported_attributes": {}
          },
          "0x0702": {
            "endpoint_attribute": "smartenergy_metering",
            "attributes": {
              "0x0000": {
                "attribute_name": "current_summ_delivered",
                "value": 17
              },
              "0x0306": {
                "attribute_name": "metering_device_type",
                "value": 0
              },
              "0x0200": {
                "attribute_name": "status",
                "value": 0
              },
              "0x0303": {
                "attribute_name": "summation_formatting",
                "value": 42
              },
              "0x0300": {
                "attribute_name": "unit_of_measure",
                "value": 0
              }
            },
            "unsupported_attributes": {
              "0x0400": {
                "attribute_name": "instantaneous_demand"
              },
              "0x0100": {
                "attribute_name": "current_tier1_summ_delivered"
              },
              "0x0302": {
                "attribute_name": "divisor"
              },
              "0x0301": {
                "attribute_name": "multiplier"
              },
              "0x0304": {
                "attribute_name": "demand_formatting"
              },
              "0x0102": {
                "attribute_name": "current_tier2_summ_delivered"
              },
              "0x0104": {
                "attribute_name": "current_tier3_summ_delivered"
              },
              "0x0106": {
                "attribute_name": "current_tier4_summ_delivered"
              },
              "0x0108": {
                "attribute_name": "current_tier5_summ_delivered"
              },
              "0x010a": {
                "attribute_name": "current_tier6_summ_delivered"
              }
            }
          },
          "0x0b04": {
            "endpoint_attribute": "electrical_measurement",
            "attributes": {
              "0x0603": {
                "attribute_name": "ac_current_divisor",
                "value": 1000
              },
              "0x0602": {
                "attribute_name": "ac_current_multiplier",
                "value": 1
              },
              "0x0605": {
                "attribute_name": "ac_power_divisor",
                "value": 1
              },
              "0x0604": {
                "attribute_name": "ac_power_multiplier",
                "value": 1
              },
              "0x0601": {
                "attribute_name": "ac_voltage_divisor",
                "value": 1
              },
              "0x0600": {
                "attribute_name": "ac_voltage_multiplier",
                "value": 1
              },
              "0x050b": {
                "attribute_name": "active_power",
                "value": 2
              },
              "0x0000": {
                "attribute_name": "measurement_type",
                "value": 3221762439
              },
              "0x0508": {
                "attribute_name": "rms_current",
                "value": 14
              },
              "0x0505": {
                "attribute_name": "rms_voltage",
                "value": 236
              }
            },
            "unsupported_attributes": {
              "0x0300": {
                "attribute_name": "ac_frequency"
              },
              "0x0401": {
                "attribute_name": "ac_frequency_divisor"
              },
              "0x0302": {
                "attribute_name": "ac_frequency_max"
              },
              "0x0400": {
                "attribute_name": "ac_frequency_multiplier"
              },
              "0x0403": {
                "attribute_name": "power_divisor"
              },
              "0x0402": {
                "attribute_name": "power_multiplier"
              },
              "0x0507": {
                "attribute_name": "rms_voltage_max"
              },
              "0x050a": {
                "attribute_name": "rms_current_max"
              },
              "0x050d": {
                "attribute_name": "active_power_max"
              },
              "0x050f": {
                "attribute_name": "apparent_power"
              },
              "0x0510": {
                "attribute_name": "power_factor"
              }
            }
          },
          "0xe001": {
            "endpoint_attribute": null,
            "attributes": {},
            "unsupported_attributes": {}
          }
        },
        "out_clusters": {
          "0x000a": {
            "endpoint_attribute": "time",
            "attributes": {},
            "unsupported_attributes": {}
          },
          "0x0019": {
            "endpoint_attribute": "ota",
            "attributes": {},
            "unsupported_attributes": {}
          }
        }
      },
      "242": {
        "device_type": {
          "name": "PROXY_BASIC",
          "id": 97
        },
        "profile_id": 41440,
        "in_clusters": {},
        "out_clusters": {
          "0x0021": {
            "endpoint_attribute": "green_power",
            "attributes": {},
            "unsupported_attributes": {}
          }
        }
      }
    }
  }
}
```

</details>

### Logs

<details><summary>Logs</summary>

```python
[Paste the logs here]
```

</details>

**Comment by TheJulianJES** [comment]
This might really be a ZHA issue. If the measurement type is kWh (0), we do not use the `summation_formatting` attribute for some reason. We only use the multiplier and divider attributes in that case: [zha/sensor.py#L522-L531](https://github.com/home-assistant/core/blob/2c196baa7a548291b72283d6acf0d8e0090e545a/homeassistant/components/zha/sensor.py#L522-L531)

@Adminiuga If you're around, do you have any idea why you added [that check](https://github.com/home-assistant/core/blob/2c196baa7a548291b72283d6acf0d8e0090e545a/homeassistant/components/zha/sensor.py#L524-L525) in the [original PR](https://github.com/home-assistant/core/pull/56666)? I'm guessing that it's unexpected the formatter is still needed for kWh? Or did some devices perhaps provide something wrong?

---

This can be worked around with a quirk. I'm not sure if that's a good idea though, as the plug seems to provide `summation_formatting` correctly: `00101010` =>  2 digits to right of decimal point, 5 digits to left of decimal point.
Changing the ZHA code could break some other devices though. Especially the ones we have quirks for which set the multiplier / dividers right now.

.. but apparently we already use quirks for other Innr plugs though: https://github.com/zigpy/zha-device-handlers/blob/dev/zhaquirks/innr/innr_sp120_plug.py
I guess I'll get one ready for this plug too for now.

**Comment by TheJulianJES** [comment]
@rrozema Please test [this file](https://raw.githubusercontent.com/zigpy/zha-device-handlers/fba96b883340a5e55926714d474002cf00eee1e9/zhaquirks/innr/innr_sp240_plug.py) as a [custom quirk](https://github.com/zigpy/zha-device-handlers/discussions/693#discussioncomment-857274).
The comment above is still relevant for the isuse though.

**Comment by tube0013** [comment]
I think the US version of this plug the 234 may have the same issue. I have my LumenPNP on one and it's like a 3d printer without heaters and some small vacuum pumps when running.  I don't think it's put out 50 kWh during it's lifetime. 

<img width="296" alt="image" src="https://github.com/zigpy/zha-device-handlers/assets/11084412/93eb0650-9997-4df5-9bb2-ce7552a6b8e8">

Looks like the current quirk is just changing the voltage divisor: https://github.com/zigpy/zha-device-handlers/blob/d30dac2f3ffe99cc303690afb91513cb8836acc4/zhaquirks/innr/innr_sp234_plug.py#L34 I

If I query the Summation_Formatting attribute I get 4 (0x2A) as well.

I can try and add the fix you added for the 240 in the quirk for the 234 to see if the work around works as well.



```
{
  "node_descriptor": "NodeDescriptor(logical_type=<LogicalType.Router: 1>, complex_descriptor_available=0, user_descriptor_available=0, reserved=0, aps_flags=0, frequency_band=<FrequencyBand.Freq2400MHz: 8>, mac_capability_flags=<MACCapabilityFlags.FullFunctionDevice|MainsPowered|RxOnWhenIdle|AllocateAddress: 142>, manufacturer_code=4454, maximum_buffer_size=82, maximum_incoming_transfer_size=82, server_mask=11264, maximum_outgoing_transfer_size=82, descriptor_capability_field=<DescriptorCapability.NONE: 0>, *allocate_address=True, *is_alternate_pan_coordinator=False, *is_coordinator=False, *is_end_device=False, *is_full_function_device=True, *is_mains_powered=True, *is_receiver_on_when_idle=True, *is_router=True, *is_security_capable=False)",
  "endpoints": {
    "1": {
      "profile_id": "0x0104",
      "device_type": "0x010a",
      "input_clusters": [
        "0x0000",
        "0x0003",
        "0x0004",
        "0x0005",
        "0x0006",
        "0x0702",
        "0x0b04",
        "0x0b05",
        "0x1000",
        "0xfc57",
        "0xfc82"
      ],
      "output_clusters": [
        "0x0003",
        "0x000a",
        "0x0019"
      ]
    },
    "242": {
      "profile_id": "0xa1e0",
      "device_type": "0x0061",
      "input_clusters": [],
      "output_clusters": [
        "0x0021"
      ]
    }
  },
  "manufacturer": "innr",
  "model": "SP 234",
  "class": "zhaquirks.innr.innr_sp234_plug.SP234"
}
```

**Comment by TheJulianJES** [comment]
@tube0013 I've pushed some changes to:
- https://github.com/zigpy/zha-device-handlers/pull/2782

To install the SP234 quirk easily as a custom quirk, use this:
<details>
<summary>SP234 custom quirk test code (click to exapnd)</summary>

```python
"""Innr SP 234 plug."""
from zigpy.profiles import zgp, zha
from zigpy.quirks import CustomCluster, CustomDevice
from zigpy.zcl.clusters.general import (
    Basic,
    GreenPowerProxy,
    Groups,
    Identify,
    OnOff,
    Ota,
    Scenes,
    Time,
)
from zigpy.zcl.clusters.homeautomation import Diagnostic, ElectricalMeasurement
from zigpy.zcl.clusters.lightlink import LightLink
from zigpy.zcl.clusters.smartenergy import Metering

from zhaquirks.const import (
    DEVICE_TYPE,
    ENDPOINTS,
    INPUT_CLUSTERS,
    MODELS_INFO,
    OUTPUT_CLUSTERS,
    PROFILE_ID,
)


class ElectricalMeasurementCluster(CustomCluster, ElectricalMeasurement):
    """Fix divisor for AC power."""

    _CONSTANT_ATTRIBUTES = {ElectricalMeasurement.AttributeDefs.ac_power_divisor.id: 1}


class MeteringClusterInnr(CustomCluster, Metering):
    """Provide constant multiplier and divisor.

    The device supplies the summation_formatting attribute correctly, but ZHA doesn't use it for kWh at the moment.
    """

    _CONSTANT_ATTRIBUTES = {
        Metering.AttributeDefs.multiplier.id: 1,
        Metering.AttributeDefs.divisor.id: 100,
    }


class SP234(CustomDevice):
    """Innr SP 234 smart plug."""

    signature = {
        MODELS_INFO: [("innr", "SP 234")],
        ENDPOINTS: {
            1: {
                PROFILE_ID: zha.PROFILE_ID,
                DEVICE_TYPE: zha.DeviceType.ON_OFF_PLUG_IN_UNIT,
                INPUT_CLUSTERS: [
                    Basic.cluster_id,
                    Identify.cluster_id,
                    Groups.cluster_id,
                    Scenes.cluster_id,
                    OnOff.cluster_id,
                    Metering.cluster_id,
                    ElectricalMeasurement.cluster_id,
                    Diagnostic.cluster_id,
                    LightLink.cluster_id,
                    0xFC57,
                    0xFC82,
                ],
                OUTPUT_CLUSTERS: [
                    Identify.cluster_id,
                    Time.cluster_id,
                    Ota.cluster_id,
                ],
            },
            242: {
                PROFILE_ID: zgp.PROFILE_ID,
                DEVICE_TYPE: zgp.DeviceType.PROXY_BASIC,
                INPUT_CLUSTERS: [],
                OUTPUT_CLUSTERS: [GreenPowerProxy.cluster_id],
            },
        },
    }

    replacement = {
        ENDPOINTS: {
            1: {
                PROFILE_ID: zha.PROFILE_ID,
                DEVICE_TYPE: zha.DeviceType.ON_OFF_PLUG_IN_UNIT,
                INPUT_CLUSTERS: [
                    Basic.cluster_id,
                    Identify.cluster_id,
                    Groups.cluster_id,
                    Scenes.cluster_id,
                    OnOff.cluster_id,
                    MeteringClusterInnr,
                    ElectricalMeasurementCluster,
                    Diagnostic.cluster_id,
                    LightLink.cluster_id,
                    0xFC57,
                    0xFC82,
                ],
                OUTPUT_CLUSTERS: [
                    Identify.cluster_id,
                    Time.cluster_id,
                    Ota.cluster_id,
                ],
            },
            242: {
                PROFILE_ID: zgp.PROFILE_ID,
                DEVICE_TYPE: zgp.DeviceType.PROXY_BASIC,
                INPUT_CLUSTERS: [],
                OUTPUT_CLUSTERS: [GreenPowerProxy.cluster_id],
            },
        },
    }
```

</details>


Let me know if all your plug measurements are correct then.

**Comment by tube0013** [comment]
Thanks got it running now.. and reset the summation delivered to get a fresh summation with the updated quirk.

**Comment by TheJulianJES** [comment]
Z2M also adds it for your plug ([code here](https://github.com/Koenkk/zigbee-herdsman-converters/blob/master/src/devices/innr.ts)). They are doing some other things we're not doing too.
Can you check if the current also needs to be divided by 1000 for your plug, @tube0013?

Actually, can you upload a diagnostics file for your plug? I'm interested in what `ac_current_divisor` returns on your plug.

**Comment by tube0013** [comment]
Amperage seems okay to me:

![IMG_3338](https://github.com/zigpy/zha-device-handlers/assets/11084412/d0495b15-bc30-4fab-998b-645d526cd24a)

ac_current_divisor is 1000:

[zha-e29ceebf7c284fe6a18d61622537f595-innr SP 234-bd4503f7c2ae9e99b0befd345d61ad65.json.txt](https://github.com/zigpy/zha-device-handlers/files/13495963/zha-e29ceebf7c284fe6a18d61622537f595-innr.SP.234-bd4503f7c2ae9e99b0befd345d61ad65.json.txt)

**Comment by rrozema** [comment]
![image](https://github.com/zigpy/zha-device-handlers/assets/1747982/ef5768eb-8c58-4ac8-a465-e72fdd55d36e)
I confirm the quirk fixes my kWh sensor.

**Comment by GerryInnr** [comment]
Good to see the issue is resolved.
Just for info: the SP 234 US plug had an issue in the first firmware with an incorrect divider that was fixed in a firmware update. That update also fixes a number of other issues and is available on https://github.com/Koenkk/zigbee-OTA/tree/master/images/Innr.

**Comment by tube0013** [comment]
SP 234 quirk changes have been working well.

Interesting I missed that fact that innr had fw updates posted. so I guess the quick need to be adjusted for that... I'm pulling it down and will update. @TheJulianJES should we move the SP 234 to a new issue sorry fop combining it with the SP 240!

**Comment by TheJulianJES** [comment]
Ah, the info with the older firmware versions is interesting. It explains why Z2M still sets those dividers manually.

https://github.com/zigpy/zha-device-handlers/pull/2782 will be included in 2023.12.0b0 and should fix the issue even for older firmware versions of the SP 234, as it always sets the EM current and EM power divisors for the SP 234 now.
The total sum delivered (Metering) fix is also included for all Innr plugs (setting divisor, as ZHA ignores `summation_formatting` at the moment).

So, the behavior should be fixed in all upcoming 2023.12.x versions.
If there any remaining issues in 2023.12.x (beta should be out in about one to two hours or so), please open a new issue for other plugs and tag me.

**Comment by tube0013** [comment]
My SP 234 was already on latest fw, so everything seems working with the modified one as I have been testing over last day or so.

**Comment by thecoop** [comment]
This is also an issue for the SP 242

**Comment by chonkusZZ** [comment]
> This is also an issue for the SP 242

I too am having the same issue with SP 242, FW 1.4.3.

I can't seem to even find this device on Innr's website anymore despite only recently purchasing them, but am exhibiting the same issue with the summation being much higher than it should be, assuming it is the same problem.

**Comment by GerryInnr** [comment]
The new plugs SP 240, SP 242, and SP 244 all use the same firmware code base so the same quirk can be applied.
Just for completeness: it is not a bug of the smart plugs; the plugs correctly provide the SummationFormatting attribute 0x0303 with value 0x2A (meaning 2 digits to the right 5 to the left of the decimal point), which is the only mandatory attribute for formatting the CurrentSummationDelivered attribute. The Multiplier and Divisor attributes 0x0301 and 0x0302 are optional.

**Comment by dmulcahey** [comment]
> The new plugs SP 240, SP 242, and SP 244 all use the same firmware code base so the same quirk can be applied. Just for completeness: it is not a bug of the smart plugs; the plugs correctly provide the SummationFormatting attribute 0x0303 with value 0x2A (meaning 2 digits to the right 5 to the left of the decimal point), which is the only mandatory attribute for formatting the CurrentSummationDelivered attribute. The Multiplier and Divisor attributes 0x0301 and 0x0302 are optional.

Hrm we should look at the HA side of this to make sure we are doing things correctly then…

**Comment by TheJulianJES** [comment]
> Hrm we should look at the HA side of this to make sure we are doing things correctly then…

Yeah, I think there's an issue with ZHA at the moment, see this comment: https://github.com/zigpy/zha-device-handlers/issues/2781#issuecomment-1830832448 (formatter isn't used for kWh)
Even if we fix ZHA code, we risk breaking multiple devices with existing quirks. Some devices report it correctly and others not, so we don't know which quirks should stay and which should be removed (without knowing the attributes reported by the device).

**Comment by exalyte** [comment]
Sorry to add to an old bug but it seems the newer innr 242 has the same bug summation is 100x expected 

Firmware reported by zha is 0x14033682 

Please let me know if this should it's own ticket

EXA

**Comment by GerryInnr** [comment]
> Sorry to add to an old bug but it seems the newer innr 242 has the same bug summation is 100x expected
> Firmware reported by zha is 0x14033682
> Please let me know if this should it's own ticket
> EXA

As described before in this thread, this is not a bug of the SP 242, it is a bug of ZHA not using the SummationFormatting attribute of the Metering cluster to determine the right scale of the CurrentSummationDelivered attribute.

**Comment by epd5** [comment]
Loading the SP 240 .py for my SP 242 worked perfectly. Thanks for providing this.

**Comment by UltimateGG** [comment]
Having this on SP 244. Can we get the quirk to match those newer models too? I assume it's just not matching because the model is different

**Comment by pajtaz** [comment]
This is still an issue with SP 240 with latest firmware 1.9.30. I tried the custom quirk but it is not working. However now the value is 10 times larger rather than 100 times.

**Comment by IWR69** [comment]
Also still an issue with SP 242 running 1.9.30

**Comment by josephnfoley** [comment]
I was struggling with this all day as I have 4 of these plugs. I am using Zigbee2Mqtt which may or may not help but in my case I have found that on the previous firmware 1.7.23, the summaFormatting was set to 42 which is aparently correct however the multiplier and divisor attributes were not configured and when I tried to set their values I got an error saying they were not supported. When I pushed an OTA update to 1.9.3 I found a few things. First the summaFormatting is now set to 51 which shows that they have changed something in the energy formatting, also the multiplier and divisor are now there and set to 1 and 1000 respectively though these are read only. Also the current now reads 1000x what it should be. To resolve the current issue I went to the Settings (Specific) tab and set the Current Calibration to 0.001 which correctly calculates the current draw. I also set the Energy Precision to 2 to give me 2 decimal places. Now everything seems to be resolved for now.

TLDR: Pushed OTA update to plug to Version 1.9.3, set Current Calibration to 0.001 and Energy Precision to 2.

**Comment by TheJulianJES** [comment]
### For everyone that wants to help fix this, here's some testing you can do:

The first custom quirk below should disable all built-in quirks for the three Innr plugs. You'll need to install the custom quirk, restart HA, then completely remove the device from HA and re-pair it, in order for ZHA to re-read all values that might have changed due to the quirk or firmware updates.

After you have done that, please provide do the following:
1. Make sure you removed the device from HA and that it was re-paired whilst the custom quirk below is installed.
2. Check if all sensors show correct values (especially the "summation delivered" sensor).
3. a) Check what firmware version you're running by clicking on the three dots on the device page, then clicking "Manage Zigbee device", staying on the Clusters tab, and selecting `Basic` in the first dropdown, then `sw_build_id` in the second one. Finally, press "Read Zigbee attribute" and note the output, though reading the attribute will automatically include it in diagnostics later.
b) Please also read the `current_file_version` attribute on the `OTA` cluster using the same process from above.
4. **Download diagnostics** from the device page -> three dots -> Download diagnostics.
5. Drag and drop the downloaded diagnostics file into the comment field in this issue.
6. In that comment, **mention if the sensors show the expected values**. Also list the firmware you're running if you noted it earlier.

### Custom quirk that relies on firmware-set values and to provide diagnostics from

Use this quirk to gather and provide diagnostics to us. Be sure to remove and re-pair the device whilst this custom quirk is installed to make sure the diagnostics only contain the values from the device.

<details><summary>1. CUSTOM QUIRK: relies on firmware values (click to open)</summary>
<p>

```python
"""Innr SP 24x plug custom quirk to override/disable built-in ones."""

from zigpy.quirks.v2 import QuirkBuilder

from zhaquirks.innr import INNR

(
    QuirkBuilder(INNR, "SP 240")
    .applies_to(INNR, "SP 242")
    .applies_to(INNR, "SP 244")
    .add_to_registry()
)
```

</p>
</details> 

The diagnostics especially should help us to identify all possible firmware variants and what multipliers/dividers we need to set for them. There's not much data on this yet, since Z2M just requires the latest firmware version for these devices.

Also note that the custom quirk above is not supposed to be installed in the long term. We just want all the raw attributes from the device, so we can develop quirks that fix invalid issues for all firmware versions.

## Testing different divider values

If you're more adventurous, you can also play with the values for the "summation delivered" entity yourself by using the quirk below and changing the `100` set for the `divisor`. You should just need to restart Home Assistant to apply the changes, no need to removed and re-pair device again.

### Custom quirk for testing custom divider values

Do **not** use this custom quirk to provide diagnostics for us here. They'll contain your modified values, instead of the ones set on the device itself.

Use this quirk only to play with the divider values yourself. However, do mention what values work for your plug model and what firmware version you're running. Read `sw_build_id` on the `Basic` cluster and `current_file_version` on the `OTA` cluster to get the updated values. Provide those in the same comment with what divider value works for you.

<details><summary>2. CUSTOM QUIRK: testing divider values yourself (click to open)</summary>
<p>

```python
"""Innr SP 24x plug custom quirk to override built-in ones."""

from zigpy.quirks import CustomCluster
from zigpy.quirks.v2 import QuirkBuilder
from zigpy.zcl.clusters.smartenergy import Metering

from zhaquirks.innr import INNR


class MeteringClusterTest(CustomCluster, Metering):
    """Provide constant multiplier and divisor for testing Innr plugs."""

    _CONSTANT_ATTRIBUTES = {
        Metering.AttributeDefs.multiplier.id: 1,
        Metering.AttributeDefs.divisor.id: 100,  # insert your own values
    }


(
    QuirkBuilder(INNR, "SP 240")
    .applies_to(INNR, "SP 242")
    .applies_to(INNR, "SP 244")
    .replaces(MeteringClusterTest)
    .add_to_registry()
)
```

</p>
</details>

**Comment by thecoop** [comment]
SP 242
`sw_build_id`: 1.4.3
`current_build_version`: 335754882

Summation delivered was just fixed at 0.000kWh with the diagnostic quirk
Setting a divisor of 100 shows the correct value

**Comment by bazcurtis** [comment]
I have two plugs. One is a few months old than the other, but the output below is the same.

sw_build_id: 1.7.23
current_build_version: 387397253

The plug that is controlling a desk fan showed 80.000 kWh in a day. The fan was 11 hours. I have not installed the quirk yet. Do you want me to install it and report back? The screenshot below is the plug running the desk fan.

<img width="555" height="456" alt="Image" src="https://github.com/user-attachments/assets/57cc1339-1b31-43a8-8d5b-ac80e24e82ac" />

**Comment by renatoyamane** [comment]
BTW, this is the suggestion from INNR:

1. In Home Assistant, go to settings > addons > addon store and install the "file editor" there.
2. Once the file editor is installed, open it and select "browse filesystem." From the list of options, select "configuration.yaml."
3. Once the configuration.yaml file is open, paste the following code into the bottom of the file and save the file:

```
zha:
  zigpy_config:
    ota:
      extra_providers:
        - type: z2m
```

4. Once the configuration.yaml changes have been made, go back to settings and click the top right corner to restart Home Assistant. Once this is done, you can click the same spot in the top right corner to have the system search for updates. The update for the plug should now appear at the top of the settings screen.

~~5. Once the update is complete, you can go to the SP 240 plug in Home Assistant and click Reconfigure under "Device Info." The plug is now updated and will work perfectly!~~

5. Once the firmware update is complete, **delete the INNR Smart Plug from your ZigBee network and re-pair it again**

**Comment by bazcurtis** [comment]
The Innr "fix" mentioned above made mine even worse. I am not sure how to reverse it. I probably should have tested one of them and not done both at the same time 😀

You can see the before and after in these screenshots.

<img width="1082" height="1394" alt="Image" src="https://github.com/user-attachments/assets/534b7ca7-9505-4045-9354-18ae66e38044" />
<img width="1060" height="1409" alt="Image" src="https://github.com/user-attachments/assets/1f7e3903-53ba-4c94-8899-1083a4729efc" />
<img width="556" height="538" alt="Image" src="https://github.com/user-attachments/assets/1928f542-b397-4978-b9bd-56a6569fb372" />
<img width="555" height="536" alt="Image" src="https://github.com/user-attachments/assets/59122762-2cc1-4cd2-b256-f08aff7df7e1" />

**Comment by puddly** [comment]
@bazcurtis can you upload diagnostics for the device after the update?

**Comment by renatoyamane** [comment]
My diagnostics after the firmware update is attached here.

The diagnostics before the update can be found on[ my bug report](https://github.com/home-assistant/core/issues/151674).

[zha-01JY94TZJP8M8Y8RT44B94DNG1-innr SP 242-5a60f4c19098e3e979f5739dc18f8f6e.json](https://github.com/user-attachments/files/22198758/zha-01JY94TZJP8M8Y8RT44B94DNG1-innr.SP.242-5a60f4c19098e3e979f5739dc18f8f6e.json)

**Comment by bazcurtis** [comment]
> [@bazcurtis](https://github.com/bazcurtis) can you upload diagnostics for the device after the update?

I am very happy to do that. Do I need to install the quirk first? If I need to remove them and re-add them, that will take a bit of time, but happy to do it.

**Comment by renatoyamane** [comment]
> [@bazcurtis](https://github.com/bazcurtis) The Innr "fix" mentioned above made mine even worse

Could you please delete it from your ZigBee network and repair it again after the firmware update?

INNR suggested to do this, after I reported that it became even worse and now it's reporting the Summation Delivered correctly.

I updated the Step 5 on the [solution](https://github.com/zigpy/zha-device-handlers/issues/2781#issuecomment-3258792532) that worked for me.

**Comment by Byh0ki** [comment]
Hello @TheJulianJES ,

I've setup the diagnostic quirk on 2 devices with different versions, here are the info for both of them:

```
SP-240_1:
  sw_build_id: 1.9.34
  FW: 0x19223685
  current_file_version: 421672581
```
Diagnostic file: [zha-01JYF9S0ZD4W66030GEFDNEADM-innr SP 240-b757a5773128d03ebe54222e624d5516.json](https://github.com/user-attachments/files/22604752/zha-01JYF9S0ZD4W66030GEFDNEADM-innr.SP.240-b757a5773128d03ebe54222e624d5516.json)

```
SP-240_2:
  sw_build_id: 1.9.30
  FW: 0x191e3685
  current_file_version: 421410437
```

Diagnostic file: [zha-01JYF9S0ZD4W66030GEFDNEADM-innr SP 240-dcd8deb59750f28160ffbb010b23e905.json](https://github.com/user-attachments/files/22604762/zha-01JYF9S0ZD4W66030GEFDNEADM-innr.SP.240-dcd8deb59750f28160ffbb010b23e905.json)

From what I can see, the reported values are good for both of the sockets. So it would confirm that the version >= 1.9.30 no longer need the quirks to report the proper values.

**Comment by AllardKatan** [comment]
I had noticed for a while that my Innr SP240 plugs were reporting too high energy values, too much to be true, while the power readings made sense with what my total home energy was reporting. I first thought this was due to the integration, and actually set up an integral sensor on the reported power to fix this. By comparing the integral sensor and the reported 'summation delivered' I found out that there is exactly a factor 10 difference by downloading the history of both entities and plotting them together.  My firmware version is 0x191d3685 accordign to the device info page, or 1.9.29 according to sw_build_id cluster attribute . If I check the 'MeteringClusterInnr' and read the 'summation_formatting' attribute, I get the following:
`NumberFormatting<0x33>(num_digits_right_of_decimal=3, num_digits_left_of_decimal=6, suppress_leading_zeros=0`

I also did debug logging of the ZHA events. I found the following lines: 
```2025-10-22 16:58:29.517 DEBUG (Thread-4) [aiosqlite] executing functools.partial(<built-in method execute of sqlite3.Connection object at 0x7fe208d7aa70>, '\n            INSERT INTO attributes_cache_v13\n            VALUES (:ieee, :endpoint_id, :cluster_type, :cluster_id, :attr_id, :value, :timestamp)\n                ON CONFLICT (ieee, endpoint_id, cluster_type, cluster_id, attr_id) DO UPDATE\n                SET value=excluded.value, last_updated=excluded.last_updated\n                WHERE\n                    value != excluded.value\n                    OR :timestamp - last_updated > :min_update_delta\n            ', {'ieee': a4:c1:38:57:0f:ad:1a:f4, 'endpoint_id': 1, 'cluster_type': <ClusterType.Server: 0>, 'cluster_id': 1794, 'attr_id': 0, 'value': 111770, 'timestamp': 1761145109.500436, 'min_update_delta': 30.0}) ```

The 'value'of 111770 is translated to HA as 1,117.700 kWh for this plug, as visible on the device info page. That is not consistent with the number formatting attribute, which states that there should be three digits to the right of the decimal, so .770. The 1,117.7 value is a factor 10 too high, I am sure of that. So whatever is happening to translate the value to kWh, it needs a fix.

(I read the post by The JulianES above a bit better now, forgot to add the current file version fromthe OTA cluster. It is 421344901 . I did not install the custom quirk, but I think I found the value for the divider without it)

**Comment by tlinnet** [comment]
Hi,
I recently bought 4x SP 240 by innr, to track energy consumption of Washing Machine, Dryer, Dishwasher and Water Kettle.
The example here, is a Bosch CTWK40, rated to 2000-2400 W.

Boiling water takes around 3 minutes. So I can calculate an estimation of kWh use: (2200W * 3 min) / (60 min/h) = 110 Wh = **0,110 kWh**

But the reported change in Energy = 9,84 kWh - 8,83 kWh = **1,02 kWh**

The current shows 9.3A, which fit with: I = P / U =  2200 W / 240 V = 9.1 A

<img width="1091" height="486" alt="Image" src="https://github.com/user-attachments/assets/34e7c8af-45cf-468d-aeca-f6f02a1cfdc2" />

<img width="1035" height="996" alt="Image" src="https://github.com/user-attachments/assets/2b0a06d4-5fbd-4555-9a2e-469893af01c3" />

<img width="905" height="389" alt="Image" src="https://github.com/user-attachments/assets/3b1bdcf2-427c-49c4-b863-6fa653e23909" />


This affect my Energy Dashboard.
Here API from [ha_sonnenbatterie](https://github.com/weltmeyer/ha_sonnenbatterie) reported total use of 18,7 kWh for the day, but the energy consumption from the 4x SP 240 by innr seems to be way off.

<img width="1159" height="449" alt="Image" src="https://github.com/user-attachments/assets/c857273c-84da-4d28-a632-6cb729160032" />

Which means I get these funny "Negative Untracked consumption" 

<img width="1657" height="659" alt="Image" src="https://github.com/user-attachments/assets/64d2084c-a64c-4bc1-ae8f-846d35cb3919" />

Debug info according to: https://github.com/zigpy/zha-device-handlers/issues/2781#issuecomment-3234748088
```
# Device info
SP 240 by innr
Connected via ZBT
Firmware: 0x19223685
Zigbee: Redacted
Zigbee info
Nwk: 0x478c
Device Type: Router
LQI: 172
RSSI: -68
Last seen: 2025-11-17T09:18:47
Power source: Mains
Quirk: zhaquirks.innr.innr_sp240_plug.SP240

# Debug info

Cluster: Basic: sw_build_id: 1.9.34
Cluster: Ota: current_file_version: 421672581
Cluster: MeteringClusterInnr: summation_formatting: NumberFormatting<0x33>(num_digits_right_of_decimal=3, num_digits_left_of_decimal=6, suppress_leading_zeros=0)
Cluster: MeteringClusterInnr: divisor: 100
Cluster: MeteringClusterInnr: multiplier: 1
```
Downloaded diagnostics:
[zha-01K2J9GFM7GFSF5HZ8Z6FPSH4F-innr SP 240-b814a6b437a08cd55d1ab3d0f46e7d7e.json](https://github.com/user-attachments/files/23577780/zha-01K2J9GFM7GFSF5HZ8Z6FPSH4F-innr.SP.240-b814a6b437a08cd55d1ab3d0f46e7d7e.json)

This seems related
https://github.com/zigpy/zha-device-handlers/issues/3857

**Comment by GerryInnr** [comment]
Hi guys,

The Innr plugs firmware history regarding the kWh (CurrentSummationDelivered) reading is as follows:
- FW < 1.9.27: divisor and multiplier attributes of Metering cluster not supported, CurrentSummationFormatting = 0x24 (5 digits left of decimal, 2 right).
- FW >= 1.9.27: divisor and multiplier attributes of Metering cluster supported and set to resp. 1000 and 1, CurrentSummationFormatting = 0x33 (6 digits left of decimal, 3 right).

So starting from firmware version 1.9.27 (hexadecimal 0x191b3685) reporting is in Wh (1/1000th of kWh) and divisor and multiplier are supported (they are optional in Zigbee while CurrentSummationFormatting is mandatory, which is why we didn't support the divisor and multiplier at first).

@tlinnet your plug has version 0x19223685 (1.9.34) but your debug output shows "MeteringClusterInnr: divisor: 100". That must be coming from the quirk, overriding the correct value from the plug which is 1000 (the quirk was there because the divisor was not supported yet). Hence the 10x factor of your readings in the UI. The quirk must be corrected or maybe not used for plugs with FW >= 1.9.27.

Cheers,
Gerry.

**Comment by tlinnet** [comment]
Hi @GerryInnr 

Following issue https://github.com/zigpy/zha-device-handlers/issues/3857, and then Pull Request https://github.com/zigpy/zigpy/pull/1606, I understand there is some kind of `quirks/v2`. 

That PR was merged into zigpy:dev 28/7, and I think I find it here: https://github.com/zigpy/zigpy/blob/a78dd9f90049a972527a67036651a92b892051a2/zigpy/quirks/v2/__init__.py#L489

But seems not to work? In "Download diagnostics" of ZHA, I searched for `    "versions": {` and found `"zigpy": "0.86.0",`.
Which is the latest release 3 weeks ago. And it's there.

I am new to quirks, so here are some self-learning references
* https://community.home-assistant.io/t/how-to-setup-local-zha-quirks/341226/3
* https://www.home-assistant.io/integrations/zha#configuration---yaml
* https://github.com/zigpy/zha-device-handlers/blob/dev/zhaquirks/innr/__init__.py
* https://github.com/zigpy/zha-device-handlers/blob/dev/zhaquirks/innr/innr_sp240_plug.py

So, according to [docs](https://www.home-assistant.io/integrations/zha#configuration---yaml), I should modify `configuration.yaml` and restart Home Assistant.
There is `enable_quirks` boolean (Optional, default: true), Enable quirks mode for devices where manufacturers didn’t follow specs., and `custom_quirks_path` string (Optional), Full path to a directory containing custom quirk modules that will take precedence over any built-in quirks matching a device.
There is a [guide here to: Manually add and use a not yet merged custom ZHA quirk](https://community.home-assistant.io/t/how-to-setup-local-zha-quirks/341226/5). 

So, I am tempted to either
* Understand why the PR for quirks/v2 is not working
* Override the quirk, with a new quirk
* Make a template sensor that divide by 10

**Comment by Byh0ki** [comment]
Following [TheJulienJES comment](https://github.com/zigpy/zha-device-handlers/issues/3857#issuecomment-3321850662) (and my comment right above), the work is indeed still in progress but I think that the comment from @GerryInnr will help the maintainer to properly map the quirks behavior with the INNR FW versions

**Comment by AllardKatan** [comment]
Thanks @GerryInnr for chiming in, it's good to have someone involved from the company! I'm not very familiar with quirks and Zigbee specifications but can I summarize it as follows?

- Prior 1.9.27, reporting was to two digits precision, the integer value represents 10 Wh in its smallest digit and should therefore be divided by 100 to get kWh 
- Since 1.9.27, reporting is to three digits precision, the integer value represents 1 Wh in its smallest digit and should therefore be divided by 1000 to get kWh
- In either of these cases, the CurrentSummationFormatting reports the correct formatting
- (speculative, @GerryInnr may be able to clarify) In future updates, the CurrentSummationFormatting will also be adjusted if needed

@TheJulianJES I understand from your comment above that ZHA for some reason does not use the formatting attribute? Is it possible to write a quirk in such a way that it does, instead of hard-coding divisors and multipliers which might give problems in the future? If this attribute is mandatory as Gerry says it sounds logical to fix it in ZHA rather than in a quirk, but I understand this might have unforeseen consequences.

**Comment by GerryInnr** [comment]
Hi @AllardKatan you summarized it correctly. In all our firmware versions, CurrentSummationFormatting correctly specifies the formatting of CurrentSummationDelivered, and will continue to do so in the future. We don't foresee further changes though, it will remain 6.3 digits for the foreseeable future.
While CurrentSummationFormatting is mandatory and the divisor and multiplier optional, more systems not only HA use the divisor and multiplier, and many or most smart plugs also use divisor and multiplier, probably because those are also used in the Electrical Measurement cluster for voltage, current, and power.

I'm not sure how HA works exactly but assuming that currently:

1. Unless a quirk is active that sets other values,
2. If the device does not provide a divisor and multiplier for CurrentSummationDelivered, they are assumed to be 1.

Then a good change could be:

1. Unless a quirk is active that sets other values,
2. If the device does not provide a divisor and multiplier for CurrentSummationDelivered, check CurrentSummationFormatting and set divisor and multiplier accordingly, and if the device does not provide CurrentSummationFormatting, assume divisor and multiplier 1.

With that change in HA, the quirk for SP 24x and OSP 240 smart plugs could be removed and it would work correctly for all firmware versions. Giving preference to the divisor and multiplier over CurrentSummationFormatting will very likely prevent issues for other smart plugs (which may have correct divisor and multiplier but have the mandatory CurrentSummationFormatting set to some incorrect default value).

**Comment by tlinnet** [comment]
Yeah!
One from Innr ! You will get 5 stars on Amazon then for the delivery. :)

I think I will try this fix: https://github.com/zigpy/zigpy/pull/1606
Make a quirk to filter only compatible firmware versions.

As an example, here is the quirk for the `SP 240 plug`, that only will implement the `zhaquirks.innr.innr_sp240_plug.SP240`, if the max_version=0x191E3685, sw_build_id: 1.9.30. Since I have `0x19223685 / 1.9.34`, the expected effect is to have no quirk.

* The `min_version` and `max_version` are integers representing the firmware version, minimum inclusive but maximum exclusive. 
* These are netadata to only apply the quirk if the device's firmware version matches.
* Example: max_version=0x191E3685, sw_build_id: 1.9.30

```python
"""Innr SP 240 plug."""

from zigpy.quirks.v2 import QuirkBuilder
from zigpy.quirks import CustomCluster
from zhaquirks.innr import MeteringClusterInnr


class InnrCluster(CustomCluster):
    """Innr manufacturer specific cluster."""

    cluster_id = 0xE001

(
    QuirkBuilder(manufacturer="innr", model="SP 240")
    # Firmware version `421410437` fixed the divisor and multiplier bug
    .firmware_version_filter(max_version=0x191E3685)
    .replaces(MeteringClusterInnr, endpoint_id=1)
    .replaces(InnrCluster, endpoint_id=1)
    .add_to_registry()
)
```

With [Advanced SSH & Web Terminal](https://github.com/hassio-addons/addon-ssh) 
```
cd
cd condig
nano configuration.yaml
```

Then scroll down, and with Shift+insert
```yaml
zha:
  custom_quirks_path: /config/custom_zha_quirks
```

Then "Ctrl+x, y, enter" to save.
```bash
mkdir -p custom_zha_quirks
cd custom_zha_quirks
nano innr_sp240_plug_1934.py
```

Insert the quirk code from above and save and restart HA.

After reboot, the `Quirk: zigpy.quirks.v2.CustomDeviceV2`

<img width="311" height="533" alt="Image" src="https://github.com/user-attachments/assets/28e0bae3-8257-475a-8cfc-e7d249b84929" />

```
Cluster: MeteringClusterInnr: summation_formatting: NumberFormatting<0x33>(num_digits_right_of_decimal=3, num_digits_left_of_decimal=6, suppress_leading_zeros=0)
Cluster: MeteringClusterInnr: divisor: 100
Cluster: MeteringClusterInnr: multiplier: 1
```

Then I tried to re-configure the device and read those values again. But no change happened-

Let me boil some water

**Comment by tlinnet** [comment]
I looked in https://github.com/zigpy/zha-device-handlers/blob/dev/zhaquirks/innr/__init__.py

<img width="816" height="692" alt="Image" src="https://github.com/user-attachments/assets/2e5ef11b-3c65-46d1-bac6-6a9f2a7f98b6" />

I saw `ElectricalMeasurementClusterInnr` had _CONSTANT_ATTRIBUTES: ElectricalMeasurement.AttributeDefs.ac_current_divisor.id: 1000.

So, I tried to read that value as well

<img width="542" height="538" alt="Image" src="https://github.com/user-attachments/assets/20822f81-9ff4-435e-89fe-9145e9f43e26" />

<img width="521" height="259" alt="Image" src="https://github.com/user-attachments/assets/c09731f3-655d-4487-ab05-a0f1934ce536" />

**Comment by tlinnet** [comment]
I tried instead this

```python

"""Innr SP 240 plug."""

from zigpy.quirks.v2 import QuirkBuilder
from zigpy.quirks import CustomCluster
from zigpy.zcl.clusters.smartenergy import Metering

class MeteringClusterInnr(CustomCluster, Metering):
    """Provide constant multiplier and divisor.

    The device seems to supply the summation_formatting attribute correctly, but ZHA doesn't use it for kWh for now.
    Cluster: MeteringClusterInnr: summation_formatting: NumberFormatting<0x33>
      num_digits_right_of_decimal=3, num_digits_left_of_decimal=6, suppress_leading_zeros=0
      => Divide by 1000
      See: https://github.com/zigpy/zha-device-handlers/issues/2781#issuecomment-3542058945
    """

    _CONSTANT_ATTRIBUTES = {
        Metering.AttributeDefs.multiplier.id: 1,
        Metering.AttributeDefs.divisor.id: 1000,
    }


class InnrCluster(CustomCluster):
    """Innr manufacturer specific cluster."""

    cluster_id = 0xE001

(
    QuirkBuilder(manufacturer="innr", model="SP 240")
    # Firmware version `421410437` fixed the divisor and multiplier bug
    .firmware_version_filter(min_version=0x19223685)
    .replaces(MeteringClusterInnr, endpoint_id=1)
    .replaces(InnrCluster, endpoint_id=1)
    .add_to_registry()
)
```

Restarted HA, and then get

<img width="661" height="557" alt="Image" src="https://github.com/user-attachments/assets/1276d954-a541-4f53-96fe-5a0a477ea191" />


```
Cluster: MeteringClusterInnr: summation_formatting: NumberFormatting<0x33>(num_digits_right_of_decimal=3, num_digits_left_of_decimal=6, suppress_leading_zeros=0)
Cluster: MeteringClusterInnr: divisor: 1000
Cluster: MeteringClusterInnr: multiplier: 1
```

<img width="547" height="521" alt="Image" src="https://github.com/user-attachments/assets/53f428b4-8d65-40d6-b21b-4db0c95a1d7c" />

<img width="705" height="542" alt="Image" src="https://github.com/user-attachments/assets/baf54ad4-ad12-4eb9-8f6b-c3ae77a415ae" />

So, I guess my Energy Dashboard will look more friendly tomorrow

**Comment by puddly** [comment]
Just one clarification: `summation_formatting` allows a device to self-describe how to *display* its readings. The leading and trailing zeroes aren't multipliers or divisors, they just let the device ask to show `1.2` as `1.2000` or `0001.20`. If I'm reading the issue discussion correctly, the device (in earlier firmware versions) did not have readable `multiplier` or `divisor` attributes in early firmwares. This was fixed in later firmwares.

@tlinnet if that is the case, your quirk should have `.firmware_version_filter(max_version=0x191B3685)`, I think? That would make the quirk only apply to devices running firmware earlier than 1.9.27. If it's running a later version, no quirk is required because the device supports both `multiplier` and `divisor` correctly. You may need to click the "reconfigure device" button to get ZHA to re-read them.

**Comment by AllardKatan** [comment]
@puddly if I read the ZigBee specification it says that the position of the decimal point is certainly part of the summation formatting attribute. 

> 10 . 4. 2. 2. 4 .4 Su mm ati onFo rm att in g Att r ibut e
> SummationFormatting provides a method to properly decipher the number of digits and the decimal location
> of the values found in the Summation Information Set of attributes. This attribute is to be decoded as follows:
> Bits 0 to 2: Number of Digits to the right of the Decimal Point.
> Bits 3 to 6: Number of Digits to the left of the Decimal Point.
> Bit 7: If set, suppress leading zeros.

I at least interpret this as telling you where in the integer value sent a decimal point should be introduced, not just the display formatting. Right? Or am I missing something?

**Comment by tlinnet** [comment]
> Just one clarification: `summation_formatting` allows a device to self-describe how to _display_ its readings. The leading and trailing zeroes aren't multipliers or divisors, they just let the device ask to show `1.2` as `1.2000` or `0001.20`. If I'm reading the issue discussion correctly, the device (in earlier firmware versions) did not have readable `multiplier` or `divisor` attributes in early firmwares. This was fixed in later firmwares.
> 
> [@tlinnet](https://github.com/tlinnet) if that is the case, your quirk should have `.firmware_version_filter(max_version=0x191B3685)`, I think? That would make the quirk only apply to devices running firmware earlier than 1.9.27. If it's running a later version, no quirk is required because the device supports both `multiplier` and `divisor` correctly. You may need to click the "reconfigure device" button to get ZHA to re-read them.

Yeah ! That was also my expectation from the first try. But the divisor was still reported as 100. So I did not know if the values from init.py are then used. In that case, there would be a need to override.

**Comment by puddly** [comment]
@AllardKatan Your guess is as good as mine, the spec unfortunately isn't clear with a lot of this stuff. From the older Smart Energy spec there's this table, which makes it seem like the raw integer value (the value of the attribute) is transformed via the `multiplier` and `divisor` attributes and then display formatted via the data encoded in `SummationFormatting`:

<img width="583" height="341" alt="Image" src="https://github.com/user-attachments/assets/129d1490-f765-47dd-b621-25c55f2dfc7b" />

Home Assistant doesn't support leading zeroes so we really can only use the "digits after decimal point" to hint to Core how to format the decimals (e.g. `1.0` or `1.000`).

**Comment by AllardKatan** [comment]
Oh, yes indeed with that table it looks like divisor and multiplication are the sole determinants of the final value. Thanks for the clarification, because the wording is rather confusing.

**Comment by GerryInnr** [comment]
Hi @puddly and @AllardKatan that table is also in the Zigbee Cluster Library R8 spec. The start of the section with the table says:
"The following set of attributes provides **_the ratios and formatting_** hints required to transform the received summations, consumptions, temperatures, or demands/ rates into displayable values. **_If the Multiplier and Divisor attribute values are non-zero_**, they are used in conjunction with the SummationFormatting, ConsumptionFormatting, DemandFormatting, and TemperatureFormatting attributes."

That is a strong suggestion that if the Multiplier and Divisor attribute values are zero (or not supported; remember that they are optional while CurrentSummationFormatting is mandatory), the Formatting attributes determine the ratios of the raw values. The table only explains how it works if there are non-zero Divisor and Multiplier.

Anyway, I agree that the spec is unclear on this point (and in many other places).

UPDATE:
Upon reading again I have to adjust the interpretation... The full text of that paragraph is as follows:
"The following set of attributes provides the ratios and formatting hints required to transform the received summations, consumptions, temperatures, or demands/ rates into displayable values. If the Multiplier and Divisor attribute values are non-zero, they are used in conjunction with the SummationFormatting, ConsumptionFormatting, DemandFormatting, and TemperatureFormatting attributes.
Equations required to accomplish this task are defined below:
Summation = Summation received * Multiplier / Divisor (formatted using SummationFormatting)
Consumption = Consumption received * Multiplier / Divisor (formatted using ConsumptionFormatting)
Demand = Demand received * Multiplier / Divisor (formatted using DemandFormatting)
Temperature = Temperature received * Multiplier / Divisor
_If the Multiplier and Divisor attribute values are zero, just the formatting hints defined in SummationFormatting, ConsumptionFormatting, DemandFormatting and TemperatureFormatting attributes are used._"

So the "set of attributes that provides the ratios and formatting hints" consists of all the Formatting attributes AND the Divisor and Multiplier. If the latter are 0 (or unsupported I would interpret) then the values are only formatted without changing their ratio.
That means that our firmware versions prior to 1.9.27 are not correct since they provide values representing 0.01 kWh per unit but don't define the Divisor.

**Comment by mhurk** [comment]
Seems that it is not (yet) working out of the box. There is a factor 10 difference between the reported 'Summation delivered' and the actual value in Home Assistant (2025.12.3) with Zigbee Home Automation. innr fw version 0x19223685

Also see https://github.com/zigpy/zha-device-handlers/issues/3857

**Comment by TheJulianJES** [comment]
For a status update on this issue for the SP 240, see the latest post in the other issue [here](https://github.com/zigpy/zha-device-handlers/issues/3857#issuecomment-3716999928).
TL;DR: This should be fixed in Home Assistant Core 2026.1.0, releasing later today/tomorrow.
EDIT: Some plugs may only work correctly with 2026.1.1, releasing on Friday.

EDIT (2025-01-20): Check this update here: https://github.com/zigpy/zha-device-handlers/issues/3857#issuecomment-3775862674

**Comment by jaroslavas** [comment]
Can confirm that updating to 2026.1.0 fixed `Summation delivered` value on IKEA Inspelning smart plug.
